src/metabase/search/scoring.clj

```diff
@@ -245,49 +246,73 @@
     :score  (model-score result)
     :name   "model"}])
 
-(defn- weighted-scores
-  [hit]
-  (->> hit
-       weights-and-scores
-       (filter :score)
-       (map (fn [{:keys [weight score] :as composite-score}]
-              (assoc composite-score :weighted-score (* weight score))))))
-
-(defn- accumulate-top-results
-  "Accumulator that saves the top n (defined by `search-config/max-filtered-results`) items sent to it"
-  ([] (PriorityQueue. search-config/max-filtered-results compare-score-and-result))
-  ([^PriorityQueue q]
-   (loop [acc []]
-     (if-let [x (.poll q)]
-       (recur (conj acc x))
-       acc)))
-  ([^PriorityQueue q item]
-   (if (>= (.size q) search-config/max-filtered-results)
-     (let [smallest (.peek q)]
-       (if (pos? (compare-score-and-result item smallest))
-         (doto q
-           (.poll)
-           (.offer item))
-         q))
-     (doto q
-       (.offer item)))))
+(defn- bounded-heap-accumulator
+  "A reducing function that maintains a queue of the largest items as determined by `kompare`. The queue is bounded
+  in size by `size`. Useful if you are interested in the largest `size` number of items without keeping the entire
+  collection in memory."
+  [size kompare]
+  (fn bounded-heap-acc
+    ([] (PriorityQueue. size kompare))
+    ([^PriorityQueue q]
+     (loop [acc []]
+       (if-let [x (.poll q)]
+         (recur (conj acc x))
+         acc)))
+    ([^PriorityQueue q item]
+     (if (>= (.size q) size)
+       (let [smallest (.peek q)]
+         (if (pos? (kompare item smallest))
+           (doto q
+             (.poll)
+             (.offer item))
+           q))
+       (doto q
+         (.offer item))))))
```

Comment:
This seems generally useful, should it go in `metabase.util` so we could potentially use it in other places?

Maybe package it up into a function with a name like `take-sorted` (analogous to `take`) that could return either a lazy seq or transducer like `take` would make sense

Reply:
I think in haskell there's a lazy way to sort. Using the implementation here there's no good way to do this. We obviously iterate through the entire collection and we end up with a priority queue with all of the biggest items. We could lazy-seq drain that but that seems almost useless at that point. I've added tests. I think it would be a good place for test.check but I don't think we've got that laying around. The tests are very similar to invariants that test.check excels at.